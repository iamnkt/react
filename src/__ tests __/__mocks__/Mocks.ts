export const CharmanderMockCard = {
  id: 'xy10-10',
  name: 'Charmander',
  images: {
    large: 'https://images.pokemontcg.io/xy10/10_hires.png',
  },
};

export const CharizardMockCard = {
  id: 'bw15-12',
  name: 'Charizard',
  images: {
    large: 'https://images.pokemontcg.io/bw15/12_hires.png',
  },
};

export const CharizardDetailedMockCard = {
  image: 'https://images.pokemontcg.io/bw15/12_hires.png"',
  name: 'Charizard',
  hp: '150',
  level: '12',
  types: ['fire'],
  rarity: 'rare',
};

export const CardsDataMock = {
  count: 10,
  data: [
    CharmanderMockCard,
    CharizardMockCard,
    CharizardMockCard,
    CharizardMockCard,
    CharizardMockCard,
    CharizardMockCard,
    CharizardMockCard,
    CharmanderMockCard,
    CharmanderMockCard,
    CharmanderMockCard,
  ],
  page: 1,
  pageSize: 8,
  totalCount: 10,
};

export const CardsEmptyDataMock = {
  count: 0,
  data: [],
  page: 1,
  pageSize: 8,
  totalCount: 0,
};
