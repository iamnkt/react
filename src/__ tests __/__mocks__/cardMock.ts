export const CharizardCardMock = {
  image: 'https://images.pokemontcg.io/bw15/12_hires.png"',
  name: 'Charizard',
  hp: '150',
  level: '12',
  types: ['fire'],
  rarity: 'rare',
};
