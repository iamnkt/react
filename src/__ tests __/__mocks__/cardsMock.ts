export const CharmanderMock = {
  id: 'xy10-10',
  name: 'Charmander',
  image: 'https://images.pokemontcg.io/xy10/10_hires.png"',
};

export const CharizardMock = {
  id: 'bw15-12',
  name: 'Charizard',
  image: 'https://images.pokemontcg.io/bw15/12_hires.png"',
};
