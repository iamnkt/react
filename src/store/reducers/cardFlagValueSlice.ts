import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface cardFlagValueState {
  isFetching: boolean;
}

const initialState: cardFlagValueState = {
  isFetching: false,
};

export const cardFlagValueSlice = createSlice({
  name: 'cardsFlag',
  initialState,
  reducers: {
    setIsFetching(state, action: PayloadAction<boolean>) {
      state.isFetching = action.payload;
    },
  },
});

export default cardFlagValueSlice.reducer;
