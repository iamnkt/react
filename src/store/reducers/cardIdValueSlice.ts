import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface cardIdValueState {
  id: string;
}

const initialState: cardIdValueState = {
  id: '',
};

export const cardIdValueSlice = createSlice({
  name: 'cardId',
  initialState,
  reducers: {
    setCardId(state, action: PayloadAction<string>) {
      state.id = action.payload;
    },
  },
});

export default cardIdValueSlice.reducer;
