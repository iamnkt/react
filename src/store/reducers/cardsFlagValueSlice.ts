import { createSlice, PayloadAction } from '@reduxjs/toolkit';

interface cardsFlagValueState {
  isLoading: boolean;
}

const initialState: cardsFlagValueState = {
  isLoading: false,
};

export const cardsFlagValueSlice = createSlice({
  name: 'cardsFlag',
  initialState,
  reducers: {
    setIsLoading(state, action: PayloadAction<boolean>) {
      state.isLoading = action.payload;
    },
  },
});

export default cardsFlagValueSlice.reducer;
