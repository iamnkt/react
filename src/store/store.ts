import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';
import searchValueReducer from './reducers/searchValueSlice';
import limitValueReducer from './reducers/limitValueSlice';
import pageValueReducer from './reducers/pageValueSlice';
import cardValueReducer from './reducers/cardValueSlice';
import { cardsAPI } from '../services/cardsService';
import cardsFlagValueReducer from './reducers/cardsFlagValueSlice';
import cardIdValueReducer from './reducers/cardIdValueSlice';
import cardFlagValueReducer from './reducers/cardFlagValueSlice';
import { persistReducer } from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const rootReducer = combineReducers({
  searchValueReducer,
  limitValueReducer,
  pageValueReducer,
  [cardsAPI.reducerPath]: cardsAPI.reducer,
  cardValueReducer,
  cardIdValueReducer,
  cardsFlagValueReducer,
  cardFlagValueReducer,
});

const persistConfig = {
  key: 'root',
  storage: AsyncStorage,
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: persistedReducer,
    preloadedState,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().concat(cardsAPI.middleware),
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
