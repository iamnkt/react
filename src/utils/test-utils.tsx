import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { setupListeners } from '@reduxjs/toolkit/dist/query';

import { AppStore, RootState, setupStore } from '../store/store';
import { BrowserRouter } from 'react-router-dom';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  setupListeners(store.dispatch);
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
