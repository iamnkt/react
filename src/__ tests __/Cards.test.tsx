import 'whatwg-fetch';
import { screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Cards from '../components/cards/cards';
import { CardsDataMock, CardsEmptyDataMock } from './__mocks__/Mocks';
import { renderWithProviders } from '../utils/test-utils';
import { setupStore } from '../store/store';
import { setIsLoading } from '../store/reducers/cardsFlagValueSlice';

const store = setupStore();

describe('Cards component', () => {
  it('renders correctly and should display the specified number of cards', () => {
    renderWithProviders(<Cards cards={CardsDataMock} />, { store });
    expect(screen.queryAllByTestId('card')).toHaveLength(2);
  });

  it('should be displayed component, when there are no cards', () => {
    renderWithProviders(<Cards cards={CardsEmptyDataMock} />, { store });
    expect(screen.queryByText(/No cards were found/i)).toBeInTheDocument();
  });

  it('should be displayed loader while loading the cards', () => {
    store.dispatch(setIsLoading(true));
    renderWithProviders(<Cards cards={CardsDataMock} />, { store });
    expect(screen.getByTestId('loader')).toBeInTheDocument();
  });
});
