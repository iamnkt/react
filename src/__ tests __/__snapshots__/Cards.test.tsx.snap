// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cards component renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cards__container"
      >
        <div
          class="card"
          data-testid="card"
          id="xy10-10"
        >
          <h4
            class="card__title"
            data-testid="card-title"
          >
            Charmander
          </h4>
          <img
            class="card__image"
            data-testif="card-image"
            src="https://images.pokemontcg.io/xy10/10_hires.png""
          />
        </div>
        <div
          class="card"
          data-testid="card"
          id="bw15-12"
        >
          <h4
            class="card__title"
            data-testid="card-title"
          >
            Charizard
          </h4>
          <img
            class="card__image"
            data-testif="card-image"
            src="https://images.pokemontcg.io/bw15/12_hires.png""
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cards__container"
    >
      <div
        class="card"
        data-testid="card"
        id="xy10-10"
      >
        <h4
          class="card__title"
          data-testid="card-title"
        >
          Charmander
        </h4>
        <img
          class="card__image"
          data-testif="card-image"
          src="https://images.pokemontcg.io/xy10/10_hires.png""
        />
      </div>
      <div
        class="card"
        data-testid="card"
        id="bw15-12"
      >
        <h4
          class="card__title"
          data-testid="card-title"
        >
          Charizard
        </h4>
        <img
          class="card__image"
          data-testif="card-image"
          src="https://images.pokemontcg.io/bw15/12_hires.png""
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
